### ISSUE TITLE
Private key loaded via ENGINE_load_private_key is not usable due to "pkey->pmeth_engine" becomes NULL

### ISSUE THREAD

**Issue by shindh84** [openssl/openssl#22945]
Hello.
I have a problem to use private key which is loaded using OpenSSL v3.0.12 engine.
It works well in previous versions including v3.0.11.
Below are error logs.

```
[RSA]
4076AAB67F000000:error:020000B3:rsa routines:rsa_ossl_private_encrypt:missing private key:../openssl-3.0.12/crypto/rsa/rsa_ossl.c:332:
4076AAB67F000000:error:1C880004:Provider routines:rsa_sign:RSA lib:../openssl-3.0.12/providers/implementations/signature/rsa_sig.c:588:
4076AAB67F000000:error:06880006:asn1 encoding routines:ASN1_item_sign_ctx:EVP lib:../openssl-3.0.12/crypto/asn1/a_sign.c:284:
[EC]
40E63E917F000000:error:0800007D:elliptic curve routines:ossl_ecdsa_simple_sign_sig:missing private key:../openssl-3.0.12/crypto/ec/ecdsa_ossl.c:225:
40E63E917F000000:error:06880006:asn1 encoding routines:ASN1_item_sign_ctx:EVP lib:../openssl-3.0.12/crypto/asn1/a_sign.c:284:
```

Follow commit causes the problem.
https://github.com/openssl/openssl/commit/2b74e75331a27fc89cad9c8ea6a26c70019300b5

```e->load_privkey()``` sets engine specific evp methods.
```EVP_PKEY_set1_RSA``` (introduced by above commit) calls ```EVP_PKEY_assign``` and ```EVP_PKEY_set_type```.

``` 
EVP_PKEY *ENGINE_load_private_key(ENGINE *e, const char *key_id,
                                  UI_METHOD *ui_method, void *callback_data)
{
 EVP_PKEY *pkey;
 ...
    // [Reporter] Engine sets its own EVP method which should be called instead of default OpenSSL EVP method
    pkey = e->load_privkey(e, key_id, ui_method, callback_data);
    ...
    /* We enforce check for legacy key */
    switch (EVP_PKEY_get_id(pkey)) {
    case EVP_PKEY_RSA:
        {
        RSA *rsa = EVP_PKEY_get1_RSA(pkey);
        // [Reporter] By calling below, pkey loses its EVP engine information
        EVP_PKEY_set1_RSA(pkey, rsa);		//  [Reporter] Causes problem
        RSA_free(rsa);
        }
        break;
 ...
```
 
Within ```pkey_set_type```, information about EVP methods provided by the engine is invalidated, that is, can't access (call) engine specific EVP methods.
 
```
static int pkey_set_type(EVP_PKEY *pkey, ENGINE *e, int type, const char *str,
                         int len, EVP_KEYMGMT *keymgmt)
{
	...
	// At here, pmeth_engine is reset to NULL
	# ifndef OPENSSL_NO_ENGINE
        /* If we have ENGINEs release them */
        ENGINE_finish(pkey->engine);
        pkey->engine = NULL;
        ENGINE_finish(pkey->pmeth_engine);
        pkey->pmeth_engine = NULL;
	# endif
	...
 
```
 
```EVP_DigestSignInit```
-> ```do_sigver_init```
-> ```EVP_PKEY_CTX_new_from_pkey```
-> ```int_ctx_new```
 
In above example ```pkey->pmeth_engine``` is null, so ```int_ctx_new``` calls OpenSSL default EVP methods.
Because default EVP methods can't handle engine specific pkey, it results in failure.
 
We fully understand that provider is recommend instead of engine, but there still exists various S/W components which use engine.
So kindly address above issue.

**Comment by t8m** [comment]
Possible fix in #22951 

Can you please test it?

**Comment by shindh84** [comment]
> Possible fix in #22951
> 
> Can you please test it?

I tested and checked that RSA private key is being loaded well by the engine I am using.
However, my test is limited only to loading RSA private key, so I carefully think that more various test with different algorithms might be needed.
Thank you for your fast support.

**Comment by t8m** [comment]
> The real , as is pointed in [#19965 (comment)](https://github.com/openssl/openssl/pull/19965#issuecomment-1399592137) is to improve decoder do not touch "external(foregin)" keys.

The question is what do you mean @petrovr by "external(foreign)" keys in this context. What I see that e_nss is doing simply cannot work properly because it calls d2i_PUBKEY() and it expects that key to be something "externally managed" that is simply impossible to do by default and it would break all expectations that keys are by default something managed by providers and the foreign keys are exceptions. However what I see in the current e_nss code is that the get1/set dance is being done so I do not see why we would need this incorrect workaround anymore. IMO it should be reverted. Possibly we could keep the foreign key set up as of #22951 but maybe it is unnecessary as the foreign key flag should have been already set when the engine creates the key.

**Comment by petrovr** [comment]
What is "external(foreign)" key - attributes set on structure internal to the cryptographic library ( brief details in  https://github.com/openssl/openssl/pull/19965#issuecomment-1399592137 ) in case of classical key management (pre 3.0).
Also comment is my objection to proposed and now committed correction. 

So nevertheless that internal structure has attributes set as for key that cannot be managed by provider key is imported to provider store. As expected exported key lost all those attributes.

----
E_NSS engine always will use work-around for all 3+ releases.

**Comment by petrovr** [comment]
It seems to me engine mentioned in this issue is based on PKEY functionality.

**Comment by t8m** [comment]
> It seems to me engine mentioned in this issue is based on PKEY functionality.

Yes, it is.

**Comment by Premkumarx** [comment]
Hello, 
 We expect this fix in latest release OpenSSL3.0.13. But, It is not yet fixed. Could you tell me, In which version we can expect fix for this issue.

**Comment by mattcaswell** [comment]
The fix seems to now be in #23063 which has not yet been merged although it was recently approved. Should be merged soon and be available in 3.0.14
